src/imap/mod.rs

```diff
@@ -1703,6 +1704,66 @@ async fn get_config_last_seen_uid<S: AsRef<str>>(context: &Context, folder: S) -
     }
 }
 
+/// Builds a list of sequence/uid sets. The returned sets have each no more than around 1000
+/// characters because according to https://tools.ietf.org/html/rfc2683#section-3.2.1.5
+/// command lines should not be much more than 1000 chars (servers should allow at least 8000 chars)
+fn build_sequence_sets(mut uids: Vec<u32>) -> Vec<String> {
+    uids.sort_unstable();
+
+    // first, try to find consecutive ranges:
+    let mut ranges: Vec<UidRange> = vec![];
+
+    for current in uids {
+        if let Some(last) = ranges.last_mut() {
+            if last.end + 1 == current {
+                last.end = current;
+                continue;
+            }
+        }
+
+        ranges.push(UidRange {
+            start: current,
+            end: current,
+        });
+    }
+
+    // Second, sort the uids into uid sets that are each below ~1000 characters
+    let mut result = vec![String::new()];
+    for range in ranges {
+        if let Some(last) = result.last_mut() {
+            last.push_str(&range.to_string());
+            last.push(',');
+            if last.len() > 990 {
+                result.push(String::new()); // Start a new uid set
+            }
+        }
+    }
+    for set in &mut result {
+        if set.ends_with(',') {
```

Comment:
Maybe add `,` only when you want to add a number and the string is non-empty, rather than removing stray `,` afterwards.